src/HEVM.hs

```diff
@@ -308,156 +306,123 @@ type Method = Id
-
-symExpBool :: Ctx -> When -> Exp Bool -> SBV Bool
-symExpBool ctx@(Ctx c m args store _) w e = case e of
-  And a b   -> symExpBool ctx w a .&& symExpBool ctx w b
-  Or a b    -> symExpBool ctx w a .|| symExpBool ctx w b
-  Impl a b  -> symExpBool ctx w a .=> symExpBool ctx w b
-  LE a b    -> symExpInt ctx w a .< symExpInt ctx w b
-  LEQ a b   -> symExpInt ctx w a .<= symExpInt ctx w b
-  GE a b    -> symExpInt ctx w a .> symExpInt ctx w b
-  GEQ a b   -> symExpInt ctx w a .>= symExpInt ctx w b
-  NEq a b   -> sNot (symExpBool ctx w (Eq a b))
-  Neg a     -> sNot (symExpBool ctx w a)
+
+symExp :: Ctx -> TypedExp Timed -> SMType
+symExp ctx ret = case ret of
+  ExpInt e -> SymInteger $ symExpInt ctx e
+  ExpBool e -> SymBool $ symExpBool ctx e
+  ExpBytes e -> SymBytes $ symExpBytes ctx e
+
+symExpBool :: Ctx -> Exp Timed Bool -> SBV Bool
+symExpBool ctx@(Ctx c m args store _) e = case e of
+  And a b   -> symExpBool ctx a .&& symExpBool ctx b
+  Or a b    -> symExpBool ctx a .|| symExpBool ctx b
+  Impl a b  -> symExpBool ctx a .=> symExpBool ctx b
+  LE a b    -> symExpInt ctx a .< symExpInt ctx b
+  LEQ a b   -> symExpInt ctx a .<= symExpInt ctx b
+  GE a b    -> symExpInt ctx a .> symExpInt ctx b
+  GEQ a b   -> symExpInt ctx a .>= symExpInt ctx b
+  NEq a b   -> sNot (symExpBool ctx (Eq a b))
+  Neg a     -> sNot (symExpBool ctx a)
   LitBool a -> literal a
   BoolVar a -> get (nameFromArg c m a) (catBools args)
-  TEntry a  -> get (nameFromItem m a) (catBools store')
-  ITE x y z -> ite (symExpBool ctx w x) (symExpBool ctx w y) (symExpBool ctx w z)
-  Eq (a :: Exp t) (b :: Exp t) -> case eqT @t @Integer of
-    Just Refl -> symExpInt ctx w a .== symExpInt ctx w b
-    Nothing -> case eqT @t @Bool of
-      Just Refl -> symExpBool ctx w a .== symExpBool ctx w b
-      Nothing -> case eqT @t @ByteString of
-        Just Refl -> symExpBytes ctx w a .== symExpBytes ctx w b
-        Nothing -> error "Internal Error: invalid expression type"
- where store' = case w of
-         Pre -> fst <$> store
-         Post -> snd <$> store
-
-symExpInt :: Ctx -> When -> Exp Integer -> SBV Integer
-symExpInt ctx@(Ctx c m args store environment) w e = case e of
-  Add a b   -> symExpInt ctx w a + symExpInt ctx w b
-  Sub a b   -> symExpInt ctx w a - symExpInt ctx w b
-  Mul a b   -> symExpInt ctx w a * symExpInt ctx w b
-  Div a b   -> symExpInt ctx w a `sDiv` symExpInt ctx w b
-  Mod a b   -> symExpInt ctx w a `sMod` symExpInt ctx w b
-  Exp a b   -> symExpInt ctx w a .^ symExpInt ctx w b
+  TEntry a t -> get (nameFromItem m a t) (catBools $ timeStore t store)
+  ITE x y z -> ite (symExpBool ctx x) (symExpBool ctx y) (symExpBool ctx z)
+  Eq a b -> fromMaybe (error "Internal error: invalid expression type")
+      $ [symExpBool  ctx a' .== symExpBool  ctx b' | a' <- gcast a, b' <- gcast b]
+    <|> [symExpInt   ctx a' .== symExpInt   ctx b' | a' <- gcast a, b' <- gcast b]
+    <|> [symExpBytes ctx a' .== symExpBytes ctx b' | a' <- gcast a, b' <- gcast b]
```

Comment:
very nice :sparkles:

Reply:
Hehe thank you, I'm very satisfied myself 😌